slideflow/mil/eval.py

```
@@ -425,3 +532,73 @@ def _predict_mil(
             y_pred.append(torch.nn.functional.softmax(model_out, dim=1).cpu().numpy())
     yp = np.concatenate(y_pred, axis=0)
     return yp, y_att
+
+
+def _predict_mil_activations(
+    model: Callable,
+    bags: Union[np.ndarray, List[str]],
+    attention: bool = False,
+    attention_pooling: str = 'avg',
+    use_lens: bool = False,
+    device: Optional[Any] = None
+) -> Tuple[np.ndarray, List[np.ndarray]]:
+
+    import torch
+
+    # Auto-detect device.
+    if device is None:
+        if next(model.parameters()).is_cuda:
+            log.debug("Auto device detection: using CUDA")
+            device = torch.device('cuda')
+        else:
+            log.debug("Auto device detection: using CPU")
+            device = torch.device('cpu')
+    elif isinstance(device, str):
+        log.debug(f"Using {device}")
+        device = torch.device(device)
+
+    y_pred = []
+    y_att  = []
+    hs= []
+    log.info("Generating predictions...")
+    if attention and not hasattr(model, 'calculate_attention'):
+        log.warning(
+            "Model '{}' does not have a method 'calculate_attention'. "
+            "Unable to calculate or display attention heatmaps.".format(
+                model.__class__.__name__
+            )
+        )
+        attention = False
+    for bag in bags:
+        loaded = torch.load(bag).to(device)
+        loaded = torch.unsqueeze(loaded, dim=0)
+        with torch.no_grad():
+            if use_lens:
+                lens = torch.from_numpy(np.array([loaded.shape[1]])).to(device)
+                model_args = (loaded, lens)
+            else:
+                model_args = (loaded,)
+            model_out = model(*model_args)
+            h = model.get_last_layer_activations(*model_args)#added
```

this function should be in all models, from the models that don't have it implemented yet it should raise an exception `function not implemented` or similar